### PR TITLE
PSI: jettison ItemsCache

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 1]
```diff
@@ -530,49 +530,54 @@ private fun processItemDeclarations(
...
+    val directlyDeclaredNames = HashSet<String>()
+    val processor = { e: ScopeEntry ->
+        directlyDeclaredNames += e.name
+        originalProcessor(e)
     }
 
-    val (starImports, itemImports) = scope.useItemList.chain(macroItems.useItemList)
-        .filter { it.isPublic || withPrivateImports }
-        .partition { it.isStarImport }
+    fun processItem(item: RsItemElement): Boolean {
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt]: The main trick is here: now we process items in source order, but we do postpone use item processing until the end.

**Comment by vlad20012** [comment]
`RsStructureViewTest` tests failed

**Comment by matklad** [comment]
Yeah, something fees a little off here...

image

**Comment by matklad** [comment]
bors try

**Comment by matklad** [comment]
Ok, bors r+ :)

**Pull request merged**